optaplanner-examples/src/main/java/org/optaplanner/examples/nurserostering/optional/score/NurseRosteringConstraintProvider.java

```diff
@@ -0,0 +1,480 @@
+                                        HardSoftScore.ONE_SOFT,
+                                        (employeeContractPair, shiftList) -> new EmployeeConsecutiveAssignmentStart(
+                                                employeeContractPair.getLeft(), shiftList.getItems().first())
+                                                        .getDistanceToFirstDayOfWeekend()
+                                                * employeeContractPair.getRight().getWeight());
+    }
+
+    Constraint endOnNotLastDayOfWeekend(ConstraintFactory constraintFactory) {
+        return getConsecutiveShifts(constraintFactory.from(BooleanContractLine.class)
+                .filter(minMaxContractLine -> minMaxContractLine
+                        .getContractLineType() == ContractLineType.COMPLETE_WEEKENDS &&
+                        minMaxContractLine.isEnabled()))
+                                .filter((employeeContractPair,
+                                        shiftList) -> new EmployeeConsecutiveAssignmentEnd(
+                                                employeeContractPair.getLeft(), shiftList.getItems().last())
+                                                        .isWeekendAndNotLastDayOfWeekend())
+                                .penalize("org.optaplanner.examples.nurserostering.solver", "endOnNotLastDayOfWeekend",
+                                        HardSoftScore.ONE_SOFT,
+                                        (employeeContractPair, shiftList) -> new EmployeeConsecutiveAssignmentEnd(
+                                                employeeContractPair.getLeft(), shiftList.getItems().last())
+                                                        .getDistanceToLastDayOfWeekend()
+                                                * employeeContractPair.getRight().getWeight());
+    }
+
+    // Identical shiftTypes during a weekend
+    Constraint identicalShiftTypesDuringWeekend(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(BooleanContractLine.class)
+                .filter(minMaxContractLine -> minMaxContractLine
+                        .getContractLineType() == ContractLineType.IDENTICAL_SHIFT_TYPES_DURING_WEEKEND &&
+                        minMaxContractLine.isEnabled())
+                .join(Employee.class, Joiners.equal(ContractLine::getContract, Employee::getContract))
+                .join(ShiftDate.class, Joiners.filtering((c, e, date) -> date.getDayOfWeek() == DayOfWeek.SUNDAY))
+                .join(ShiftType.class)
+                .map(QuadTuple::new)
+                .join(ShiftAssignment.class,
+                        Joiners.equal(t -> t.c.getWeekendSundayIndex(), ShiftAssignment::getWeekendSundayIndex),
+                        Joiners.equal(t -> t.b, ShiftAssignment::getEmployee),
+                        Joiners.equal(t -> t.d, ShiftAssignment::getShiftType),
+                        Joiners.filtering((t, shift) -> shift.isWeekend()))
```

Comment:
Instead of the many joins and the mapping, wouldn't it be better to just group by employee+shiftType on the shift assignment directly?

Quad joins will be very expensive on their own, we should try to avoid them as much as possible.